File: service/app/src/main/java/io/miragon/blueprint/application/port/outbound/LeasingProcess.java
package io.miragon.blueprint.application.port.outbound;

import io.miragon.blueprint.domain.bike.BikeId;
import io.miragon.blueprint.domain.leasing.ApplicationId;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import org.jspecify.annotations.Nullable;

/**
 * Outbound port that drives the BPMN engine: it starts a process instance, correlates the messages
 * that release the process's wait states, and completes the alternative-clarification user task from
 * the outside. Implemented by the Operaton adapter.
 */
public interface LeasingProcess {

    void submitRequest(LeasingApplication application);

    void correlateContractSigned(ApplicationId id);

    void correlateHandoverReported(ApplicationId id);

    void correlateApplicationWithdrawn(ApplicationId id);

    void completeAlternativeClarification(ApplicationId id, boolean alternativeFound, @Nullable BikeId bikeId);

    default void completeAlternativeClarification(ApplicationId id, boolean alternativeFound) {
        completeAlternativeClarification(id, alternativeFound, null);
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/application/port/outbound/LeasingProcess.kt
package io.miragon.blueprint.application.port.outbound

import io.miragon.blueprint.domain.leasing.ApplicationId
import io.miragon.blueprint.domain.bike.BikeId
import io.miragon.blueprint.domain.leasing.LeasingApplication

/**
 * Outbound port that drives the BPMN engine: it starts a process instance, correlates the messages
 * that release the process's wait states, and completes the alternative-clarification user task from
 * the outside. Implemented by the Operaton adapter.
 */
interface LeasingProcess {
    fun submitRequest(application: LeasingApplication)

    fun correlateContractSigned(id: ApplicationId)

    fun correlateHandoverReported(id: ApplicationId)

    fun correlateApplicationWithdrawn(id: ApplicationId)

    fun completeAlternativeClarification(
        id: ApplicationId,
        alternativeFound: Boolean,
        bikeId: BikeId? = null,
    )
}
